L(-2) Config / Command-line work

seancribbs opened the pull request:
Tasks:
- [x] Redesign configuration schema, fill out documentation
- [x] Update schema tests
- [x] Move all commands into `riak_repl_console`, simplifying `riak-repl` script
- [x] Print usage output for all commands where possible
- [x] Incorporate new 2.0.1+ settings

Required PRs:
- basho/cuttlefish#166 - adds `cuttlefish:warn/1` used in translations

Related PRs:
- basho/riak_test#686
- basho/riak_ee#284

macintux committed:
Fixes to make edoc happy

macintux committed:
More superficial cleanup: consistency, typos, fix module comments that were unintentionally attributed to the first function

macintux committed:
Fix dialyzer warnings after line adjustments

macintux committed:
Moar monospace

macintux committed:
Remember how to monospace

borshop committed:
Merge pull request #610 from basho/jrd/feature/edoc-fixes

Various fixes for edoc

Reviewed-by: reiddraper

jaredmorrow committed:
Merge branch '2.0' into develop

Commit:
Reworked riak_repl.schema

Commit:
WIP sample outputs for riak-repl script

Commit:
Fix schema errors and tests for new settings

Commit:
Add tests for fullsync interval translation

Commit:
Add heartbeat interval test covering the translation

Commit:
Missed committing lager backend in tests

Commit:
Cleanup translation with better pattern matching

Commit:
Add assertions to cover all settings

lordnull committed:
Implement soft_exit, primarily for aae_fullsyn.

Problem: transient failures of aae, such as trees not yet built or locks not
being aquired, would cause an aae fullsync process to exit abnormally. This
could happen several times in a row, creating log spam.

Resolution: the concept of soft_exit. A soft_exit is a message sent from a soon
to be exiting process to a soft_linked process. The exiting process would then
exit normally, while any soft_linked processes could handle the soft_exit
message in a similar fashion as an exit message. This would indicate an exit
reason that should be handled, but not bad enough to have the system logger
know about it.

The soft_exit message sent from the aae worker to the fscoordinator is
as simple as `{soft_exit, pid(), term()}'.

The current implementation is not generic. There can only one soft_link to
the aae, and there's no general mechanism to use soft_link's or soft_exits
elsewhere in the code base. Sorry.

Another change rolled into this is consistent use of a #partition_info record
in the fscoordinator, and error tracking the fscoordinator's state. By swapping
to useing a single data structure in the partition queue, whereis waiting list,
and purgatory queues it makes it easier to understand the fscordinator (as
there is less code modify structures).

This is a forward port of the fix done for 1.4. Conflicts favor existing code
where it does not directly effect the fix.

Conflicts:
	Makefile
	rebar.config
	src/riak_repl2_fssource.erl
	src/riak_repl2_rtq_proxy.erl
	src/riak_repl_aae_source.erl
	test/riak_core_cluster_mgr_tests.erl

Commit:
Implement new repl console UI to match new config

Commit:
Formatting fixes

Commit:
Fix output formatting

* `standard_error` group leader cannot be used because it is not
  forwarded to the RPC caller.
* Fix newlines around top-level usage and V2 warnings

Commit:
Give some output when running fullsync commands

Commit:
Fix usage output of realtime command.

Commit:
Make commands more explicit in a few cases

seancribbs commented:
Note: this is not quite finished, do not review yet.

Commit:
Add fullsync strategy to replication schema

Commit:
Add experimental schema.

These settings were extracted from the replication code, but have not
been vetted or necessarily named properly. Many of them are slated to be
removed or changed for future versions of riak_repl. To promote a
setting to the main schema, one should write documentation and tests to
cover the setting.

seancribbs commented:
Ready for review.

Commit:
Remove commands.txt stub

jonmeredith commented:
@lordnull is this something you could take?

lordnull commented on src/riak_repl_console.erl:
```diff
@@ -571,12 +892,12 @@ show_nat_map([]) ->
-            [print_ip_and_maybe_port(Int), print_ip_and_maybe_port(Ext)])
-        || {Int, Ext} <- riak_repl_ring:get_nat_map(Ring)].
+               [print_ip_and_maybe_port(Int), print_ip_and_maybe_port(Ext)])
+     || {Int, Ext} <- riak_repl_ring:get_nat_map(Ring)].
```
Everywhere else the remainder of a list comprehension is indented farther than the start, yet here it's the opposite. It's rather jarring.

seancribbs replied:
It's aligned with the opening paren of `io:format()`. Honestly this is probably just what `indent-buffer` gave me.

lordnull commented on src/riak_repl_console.erl:
```diff
@@ -502,19 +819,23 @@ modes(NewModes) ->
+    riak_core_ring_manager:ring_trans(fun riak_repl_ring:rt_cascades_trans/2, 
+                                      never),
+    io:format("Realtime cascades disabled.~n");
+realtime_cascades(["show"]) ->
```
I'm wondering what the reason behind need 'show' to show the current setting is. Clustername omits the parameter to show current setting, and this is an added change from existing behavior.

seancribbs replied:
@angrycub can speak more to this, but I think the reasoning is that there shouldn't be implicit actions. If there's any subcommands at all, there should be something filling that spot.

seancribbs replied:
This reminds me that I may have missed making the "modes" command work this way.

lordnull replied:
So it seems the difference is cluster name is assigning a value to cluster name (noun), while the cascading is taking an action, either enable, disable, or show (verb). The primary difference is cascading is being changed from setting a value (always/never) to being a verb.

I remember making the cascading setting noun oriented for future extension. However, since it's going to get a rather drastic re-design/rebuild, I don't think it should be a blocker.

lordnull commented:
With the cuttlefish and riak_ee pulls, the r_t tests pass, and manual poking of the cli interface is favorable. The eunit tests for repl do not, primarily the 'riak_repl_schema_tests' do not.

Commit:
Fix failure in schema tests

seancribbs commented:
@lordnull Fixed the 'riak_repl_schema_tests' in d980e1b, could you have another poke?

seancribbs commented:
Also, @javajolt reviewed basho/cuttlefish#166, so that should no longer cause build errors.

lordnull commented:
Eunit tests now pass, so looks good from here.

seancribbs commented:
:+1: d980e1b r=lordnull

seancribbs commented:
Ugh, failing because cuttlefish is pinned to a tag deep in bitcask. Fixing.

lordnull commented:
@borshop: retry

jonmeredith commented:
@seancribbs anything left to do on cfish/repl integration other than making unit tests pass?  release notes?

jonmeredith commented:
@borshop: retry

jonmeredith commented:
@borshop retry

Without a colon - WHY U SAY IT IS BAD?

seancribbs commented:
@jonmeredith @lordnull Do not merge this yet, I have lots of fixes for CLI inconsistencies.

lordnull committed:
Fixed invalid call to update error count track.

Increment_error_dict expects the partition, elementN of error dict, and the
state. It pulls the dict out of the state so it put it back in place, thus just
returning the state. So this call that passed the dict in was wrong.

engelsanchez committed:
Fix error/retry exit counts on location down msgs

When a partition is not available, perhaps after a number of retries,
the error exits stat should be incremented. Also, the retry exits stat
should be incremented on each retry.  This was discovered when
backporting the repl_location_failures riak_test.

engelsanchez commented:
@seancribbs what is the list of improvements that you were hoping to add here?

seancribbs commented:
@engelsanchez This is on hold until we can use riak_cli properly. A lot of the first pass of this motivated riak_cli, and I don't want to rewrite this three times.

engelsanchez commented:
Got it. Thanks!

engelsanchez committed:
Fix dialyzer warnings

The one in riak_repl2_fssource is a legit bug in the code

borshop committed:
Merge pull request #636 from basho/feature/mw/forward-port-chatty-transient-aae-fs-failures

Implement soft_exit, primarily for aae_fullsyn.

Reviewed-by: engelsanchez

engelsanchez committed:
Merge branch '2.0' into develop

Conflicts:
	dialyzer.ignore-warnings
	rebar.config

Basho-JIRA commented:
[~dsomogyi] I changed the repo/issue number back to the original. Unfortunately I can't create sub-tasks of sub-tasks, or clique#45 would have been that. 

**_[posted via JIRA by Sean Cribbs]**_

Basho-JIRA commented:
Sounds good.  Andrew made me aware of the relationship so I thought I'd update it.  I'll include this for reference.

https://github.com/basho/clique/pull/45 is needed for this issue. 

**_[posted via JIRA by Derek Somogyi]**_

lordnull commented:
@seancribbs , @engelsanchez with basho/clique#45 merged in, is this ready for a fix-up / review / merge?

cmeiklejohn committed:
Merge branch '2.0' into develop-2.0-merge

Conflicts:
	src/riak_repl2_fscoordinator.erl

Commit:
WIP splitting out v2 and v3 commands, implementing clique

Commit:
WIP Add proxy-get commands and usage output

Commit:
WIP Refactor ubiquitous text+io_lib into a local function

Commit:
Minor ws cleanup

Commit:
WIP ensure all calls to generate output using text_msg include
deprecation/upgrade warnings

Commit:
Minor formatting tweak

Commit:
Cleanup some unused functions

Commit:
Add max_fs* configs

Commit:
Refactor config max_fs* settings

borshop committed:
Merge pull request #658 from basho/develop-2.0-merge

Develop 2.0 merge.

Reviewed-by: seancribbs

cmeiklejohn committed:
Add support for Erlang 17.

borshop committed:
Merge pull request #659 from basho/17

Add support for Erlang 17.

Reviewed-by: andrewjstone

Commit:
Add nat-map commands

seancribbs commented:
Closing so this can be reopened against develop.
